pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -75,52 +72,30 @@ func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner
 		return nil, fmt.Errorf("injecting topology, %w", err)
 	}
 	// Separate pods into schedules of isomorphic scheduling constraints.
-	schedules, err := s.getSchedules(ctx, constraints, pods)
-	if err != nil {
-		return nil, fmt.Errorf("getting schedules, %w", err)
-	}
+	schedules := s.getSchedules(constraints, pods)
 	return schedules, nil
 }
 
 // getSchedules separates pods into a set of schedules. All pods in each group
 // contain isomorphic scheduling constraints and can be deployed together on the
 // same node, or multiple similar nodes if the pods exceed one node's capacity.
-func (s *Scheduler) getSchedules(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) ([]*Schedule, error) {
-	// schedule uniqueness is tracked by hash(Constraints)
-	schedules := map[uint64]*Schedule{}
+func (s *Scheduler) getSchedules(constraints *v1alpha5.Constraints, pods []*v1.Pod) []*Schedule {
+	schedules := []*Schedule{}
 	for _, pod := range pods {
-		if err := constraints.ValidatePod(pod); err != nil {
-			logging.FromContext(ctx).Infof("Unable to schedule pod %s/%s, %s", pod.Namespace, pod.Name, err)
-			continue
-		}
-		tightened := constraints.Tighten(pod)
-
-		// schedulingConstraints applies the provisioner constraints
-		// and any inferred constraints such as GPU resource requests from the pods
-		// and is then hashed to compute the schedules
-		schedulingConstraints := struct {
-			*v1alpha5.Constraints
-			GPURequests v1.ResourceList
-		}{
-			Constraints: tightened,
-			GPURequests: resources.GPULimitsFor(pod),
+		isCompatible := false
+		for index, schedule := range schedules {
+			if err := schedule.Requirements.Compatible(v1alpha5.NewPodRequirements(pod)); err == nil {
+				//TODO: The cross product of the constraints values from cloud provider may not be supported by any instance type.
+				// Need to test if there is any instance type that can support the combined constraints.
+				schedules[index].Constraints = schedules[index].Tighten(pod)
+				schedules[index].Pods = append(schedules[index].Pods, pod)
+				isCompatible = true
```

Comment:
Instead of using a boolean and break, can we simply return once we've found a match? 

This is a pretty common idiom

```
for ... {
    if found {
       return ...
    }
}
// not found case
return ...

```

Reply:
Unfortunately it is not easy to use that idiom in this case because we still need to loop through other pods after the `if found` case. The actual common idiom that we need is more like
```
for ... {
NewIteration:
    for ... {
        if found{
            continue NewIteration
        } 
        // Do something if not found
    }
}
```
I don't like it because it closely resemble `goto`. But I am open to suggestion here.